Supporting action effects and conditions in item module

jessxia opened the pull request:
This pull request includes all code from issues to support the addition of actions into the item struct. 

Issue #215: Creating the structs for game_action_t, game_action_condition_t, and game_action_effect_t and functions for adding/getting/checking possible actions for an item.

Issue #315: Developing functions for the game_action_condition_t struct, including functions to create new conditions, add them to items, and check if conditions are met

Issue #316: Developing functions for the game_action_effect_t struct, including functions to create new effects, add them to items, and carry out effects

Issue #331: Adding a hash table of items into the game struct and adding functions to add and retrieve items from the hash table. 

We had to combine 331 in a PR with 316 and 315 since the latter two rely on the item hash table implemented in issue 331. 315 and 316 were created to split the work started in 215.

jessxia commented:
Changed possible_action function back to allowed_action to resolve conflicts with AM's code. Also changed 2 tests in test_items.c that were incorrect. All tests now pass.

yoonsjeong commented on src/game-state/src/game_action.c:
```diff
@@ -0,0 +1,282 @@
+
+
+/* see game_action.h */
+bool allowed_action(item_t *item, char* action_name)
```
looking at our code, it seems our allowed_action() function checks to see if the output is SUCCESS (0) or FAILURE (1) whereas bool is True (1) and False (0). it seems like the tests pass right now, but just to prevent breakage of the code I think making this an int output might be in order

MTUC08 replied:
The new item changes use the possible action function and handle the bool output but I can change it if need be. Please let me know ASAP your final decision

MTUC08 replied:
@jessxia

jessxia replied:
Done!

yoonsjeong commented on src/game-state/src/game_action.c:
```diff
@@ -0,0 +1,282 @@
+}
+
+/* see game_action.h */
+bool check_condition(game_action_condition_t *condition)
```
I'm not sure if AM is using this (the item functions are currently being worked on @MTUC08  @SnowballAntrobus) but following allowed_action() it might help to have an int output SUCCESS/FAILURE just for consistency

MTUC08 replied:
@yoonsjeong our current implementation uses possible action and checks for bools but I can alter for consistency if need be, just let me know ASAP please @jessxia

jessxia replied:
@MTUC08 At this point I'd like to leave it if your functions already handle it as is

MTUC08 replied:
@jessxia yeah our do_item and do_item_item functions currently use possible_action and the bool format. Should I leave it as is (just to confirm)? They just aren't in master yet. If yes, could you change it back to possible_action?

MTUC08 replied:
Actually in the interest of consistency, would you mind changing it to possible_action but return SUCCESS and FAILURE instead of a bool?

jessxia replied:
@MTUC08 I can do that. Just change possible_action to bool or check_condition as well?

jessxia replied:
Will your code be pushed to master before ours? Because changing the function name will cause our build to fail

MTUC08 replied:
I think AM just uses possible_action so just that since I don't want to break anyone else's code.

MTUC08 replied:
Could you change all_conditions_met to return SUCCESS/FAILURE as well?

MTUC08 replied:
let me know if you would prefer not so I can revert my changes. Thanks!

MTUC08 replied:
Sorry just saw your other comment. You can change the function name in the action management.c file from allowed_action to possible_action on your branch and it should build right?

MTUC08 replied:
@jessxia

jessxia replied:
All conditions met should already be returning SUCCESS/FAILURE. Let me know if you need anything else to be changed!

MTUC08 replied:
oh ok awesome thanks! Did changing the function name to possible_action in actionmanagement.c fix the build?

jessxia replied:
I haven't edited any of your team's files but I will try that

jessxia commented:
Changed allowed_action back to possible_action. It now returns SUCCESS and FAILURE instead of bools. Also edited AM's actionmanagement.c to change calls of that function to match new function name.

patrickkoopmans commented on src/game-state/src/game.c:
```diff
@@ -145,17 +166,68 @@ int delete_room_llist(room_list_t *head) {
+{
+    item_t *item_src = get_item_from_game(game, item_src_name);
+    if (item_src == NULL) {
+        return 2;
```
It may be helpful to define a number of macros to return in this function as having to keep track of all these different return codes increases the cognitive load of using our interface

patrickkoopmans commented on src/game-state/src/game_action.c:
```diff
@@ -0,0 +1,282 @@
+    assert(new_action != NULL);
+
+    strncpy(new_action->action_name, act_name, strlen(act_name));
+    // new_action->action_type = act_type; //REMOVED action_type
```
Delete this line

patrickkoopmans commented on src/game-state/src/game_action.c:
```diff
@@ -0,0 +1,282 @@
+{
+    game_action_t *new_action = malloc(sizeof(game_action_t));
+    new_action->action_name = malloc(MAX_ID_LEN * sizeof(char));
+    // new_action->action_type = malloc(sizeof(action_type_t)); //REMOVED action_type
```
Another line to delete

patrickkoopmans commented on src/game-state/src/game_action.c:
```diff
@@ -0,0 +1,282 @@
+
+    if (new_action == NULL || new_action->action_name == NULL) {
+
+        fprintf(stderr,
```
As mentioned in the team meeting, we should delete all instances of stderr such as this.

patrickkoopmans commented on src/game-state/src/game_action.c:
```diff
@@ -0,0 +1,282 @@
+    game_action_t* check = get_action(item, action_name);
+    if (check != NULL) 
+    {
+        //fprintf(stderr, "Error: this action is already present.\n");
```
Delete this line

patrickkoopmans commented on src/game-state/src/game_action.c:
```diff
@@ -0,0 +1,282 @@
+
+    game_action_t *ret_action = get_action(item, action->action_name);
+    if (ret_action == NULL) {
+        return 2;
```
Another function where defined macros might help

patrickkoopmans commented on src/game-state/src/game_action.c:
```diff
@@ -0,0 +1,282 @@
+        return true;
+
+    switch(actual_attribute->attribute_tag)
+    {
```
Change the indentation to K&R

patrickkoopmans commented on src/game-state/src/game_action.c:
```diff
@@ -0,0 +1,282 @@
+{
+    //call allowed action to see if the action exists
+    if (possible_action(item, action_name))
+    {
```
Two points: macros might help, and use K&R

patrickkoopmans commented on src/game-state/src/game_action.c:
```diff
@@ -0,0 +1,282 @@
+    }
+    if(item_to_modify == NULL)
+    {
+        return 3;
```
Macros that define the return ints here would also overlap with add_action_cond()

patrickkoopmans commented on src/game-state/src/game_action.c:
```diff
@@ -0,0 +1,282 @@
+    return SUCCESS;
+}
+
+/* see game_action.h */
```
This should be an internal function and added to common_game_action.h as we want other teams to only interface with add_action_effect()

liuyifan-eric replied:
done

patrickkoopmans commented on src/game-state/src/game_action.c:
```diff
@@ -0,0 +1,282 @@
+
+
+/* see game_action.h */
+int do_all_effects(item_t *item, char* action_name) 
```
Change function indentation from Allman to K&R

patrickkoopmans commented on src/game-state/src/item.c:
```diff
@@ -230,7 +193,7 @@ int set_bool_attr(item_t* item, char* attr_name, bool value)
         new_attribute->attribute_tag = BOOLE;
         new_attribute->attribute_value.bool_val = value;
-	new_attribute->attribute_key = strndup(attr_name, 100);
+	    new_attribute->attribute_key = strndup(attr_name, 100);
```
Why is this indented?

liuyifan-eric replied:
tab -> 4 spaces.

patrickkoopmans commented on src/game-state/src/item.c:
```diff
@@ -206,7 +169,7 @@ int set_char_attr(item_t* item, char* attr_name, char value)
         new_attribute->attribute_tag = CHARACTER;
         new_attribute->attribute_value.char_val = value;
-	new_attribute->attribute_key = strndup(attr_name, 100);
+	    new_attribute->attribute_key = strndup(attr_name, 100);
```
Why is this indented?

liuyifan-eric replied:
Was a tab. Resolved.

patrickkoopmans commented on src/game-state/src/item.c:
```diff
@@ -393,80 +317,63 @@ int attributes_equal(item_t* item_1, item_t* item_2, char* attribute_name)
     }
-    int comparison = 0;
+    int comparison = FAILURE;
     switch(attribute_1->attribute_tag)
```
Allman vs K&R for this entire switch statement

patrickkoopmans commented on src/game-state/src/item.c:
```diff
@@ -393,80 +317,63 @@ int attributes_equal(item_t* item_1, item_t* item_2, char* attribute_name)
+int game_action_free(game_action_t* game_action)
+{
+    free(game_action->action_name);
+    //FREE ACTION TYPE USING AM's FUNCTION
```
Have we implemented these three in-line comments? Do they not matter? I think this is incomplete.

liuyifan-eric replied:
Do we have this "action type" in our action struct?

patrickkoopmans commented on src/game-state/include/room.h:
```diff
@@ -39,7 +20,7 @@ typedef struct path {
     char *direction; // *letter case matters*
     struct room *dest;
-    condition_list_t conditions;
+    item_t *through; //to check if the door is "open"
```
We should probably clarify that the item_t* is intended to be a door item, and that the pointer exists to check if it has an attribute "open" that is true

liuyifan-eric replied:
Have updated a more detailed explanation (comment)

patrickkoopmans commented on src/game-state/src/game.c:
```diff
@@ -43,15 +43,30 @@ int add_player_to_game(game_t *game, player_t *player) {
+
+    if (check != NULL) {
+        /* WARNING */
+        fprintf(stderr, "add_item_to_game: this item id is already in use.\n");
```
As mentioned in the team meeting we should get rid of all stderr instances.

patrickkoopmans commented on src/game-state/src/game_action.c:
```diff
@@ -0,0 +1,282 @@
+
+/* see game_action.h */
+int possible_action(item_t *item, char* action_name)
+{
```
Allman vs K&R needs to be consistent here

jessxia commented on src/game-state/include/game.h:
```diff
@@ -102,22 +108,36 @@ int add_player_to_game(game_t *game, player_t *player);
-* game, Source room_id, destination room_id, direction
+*  pointer to a game
+*  source room_id
+*  destination room_id
```
sad reac

jessxia commented on src/action_management/src/actionmanagement.c:
```diff
@@ -72,7 +72,7 @@ int do_item_action(game_t *g, action_type_t *a, item_t *i, char **ret_string)
     }
     // checks if the action can be used on the item
-    int allowed = allowed_action(i, a->c_name);
+    int allowed = possible_action(i, a->c_name);
```
ye to clarify their recent PR + ours together should have compatible code for this function

jessxia commented:
Kate noticed that in add_condition and add_effect, we were grabbing a pointer to the attribute from the wrong item. I corrected that in the most recent commit, changing item_src to item_modify for the get_attribute call within the functions.

lydiafilipe commented:
There seems to be merge conflicts with master, please fix these

jessxia commented:
Changes made to PR:

- hidden pointers in hash table typedefs across all header files removed and tests changed
- removed stderr calls in files
- merged in master (stay tuned for another commit to fix the build)

jessxia commented:
Included directories of every team in WDL and AM's Makefiles. Our build was failing due to some of ui's code not being found.

jessxia commented:
Matthew noticed that (check != NULL) within add_final_room_to_game should be (check == NULL) in game.c. Made the change.

liuyifan-eric commented on src/game-state/include/game_action.h:
```diff
@@ -0,0 +1,163 @@
+ * error 2 if action NULL
+ * error 3 if item to modify is null
+ */
+int add_action_effect(game_action_t *action, item_t *item_to_add, item_t *item_to_modify, attribute_t *attribute, attribute_value_t new_value);
```
What is the difference between item_to_add & item_to_modify? I think the latter is a typo. Remember also modify the .c file.

jessxia replied:
removed item_to_modify from this function and its call in add_effect in game.c. documentation already only has one item pointer has a parameter

liuyifan-eric commented:
added functions for freeing llists of action_effect and action_condition;
moved game_action_new() to the private common-game-action.h;
improved notes, comments, and formats

There are other improvements needed as mentioned in the above reviews.

liuyifan-eric commented on src/game-state/include/game_action.h:
```diff
@@ -0,0 +1,163 @@
+* Parameters:
+* - item to modify
+* - attribute_t
+* - attribute_value_t
```
ok

liuyifan-eric commented:
Also, macros for return values have not been added yet

liuyifan-eric commented on src/game-state/src/common-item.h:
```diff
@@ -13,8 +13,7 @@
 */
-int item_init(item_t *new_item, char *item_id, char *short_desc,
-    char *long_desc);
+int item_init(item_t *new_item, char *item_id, char *short_desc, char *long_desc);
```
updated

liuyifan-eric commented on src/game-state/include/item.h:
```diff
@@ -141,7 +166,7 @@ typedef struct attribute_wrapped_for_llist {
 int attribute_free(attribute_t *attribute);
 
-/* attributes_equal() checks if two items have the same attribute
+/* attributes_equal() checks if two items have the same attributes
```
updated

liuyifan-eric commented on src/game-state/include/game_action.h:
```diff
@@ -0,0 +1,163 @@
+ * Returns:
+ *  SUCCESS if all effects are set, FAILURE if not
+ */
+int do_all_effects(item_t* item, char* action_name); 
```
√

liuyifan-eric commented on src/game-state/include/game_action.h:
```diff
@@ -0,0 +1,163 @@
+
+/* do_effects() sets an effect
+ * Parameters:
+ * - a pointer to effect
```
okk

patrickkoopmans commented:
One thing is noticed just now while working on our test branch is that get_attribute() should be moved from common-item.h to item.h. This should be done in this PR as it pertains to the way WDL hopes to load attributes.

jessxia commented on src/game-state/src/common-game-action.h:
```diff
@@ -0,0 +1,28 @@
+*/
+game_action_t *game_action_new(char *action_name, char* success_str, char* fail_str);
+
+int do_effect(game_action_effect_t *effect);
```
added

jessxia commented:
Added documentation for do_effect, removed unnecessary item pointer input to add_action_effect, and moved get_attribute from common-item.h to item.h.

Also removed comment //FREE ACTION TYPE USING AM's FUNCTION within game_action_free since our action struct no longer contains an action_type_t

lydiafilipe commented on src/game-state/tests/test_items.c:
```diff
@@ -914,43 +854,20 @@ Test(attribute_equal, char_to_bool)
 /* Checks adding an action to an item */
 Test(attribute, add_an_action)
```
Why is this entirely blank now? is there not a way to test it? Same for the functions below

a-matai replied:
@lydiafilipe it's because we removed the functions—they were placeholders in Sprint 3. tests for the other functions are on develop_actions, not part of this PR.

jessxia commented:
Shortened long lines, removed commented out code, added in missing brackets for if statement.

jessxia commented:
Removed empty tests for obsolete action functions from test_items.c. Fixed styling to match k&r in game_action.c.

jessxia commented:
Added macros for error codes for functions in game_action.c and game.c.

lydiafilipe commented on src/game-state/src/game_action.c:
```diff
@@ -0,0 +1,273 @@
+/* see game_action.h */
+int add_action_condition(item_t *item, game_action_t *action,
+			 item_t *cond_item, attribute_t *cond_attribute, attribute_value_t cond_value) {
+    if (item == NULL) {
```
Please use the macros here

lydiafilipe commented on src/game-state/src/game_action.c:
```diff
@@ -0,0 +1,273 @@
+
+/* see game_action.h */
+//we either use item_to_add or action as action is loacted within item_to_add
+int add_action_effect(game_action_t *action, item_t *item_to_add, attribute_t *attribute, attribute_value_t new_value) {
```
Same here about macros

jessxia commented:
Done!!!!!11

jessxia commented:
Closing comments:
Game-state code to support actions, their effects, and conditions is now in master. This includes:

- Issue #215: Creating the structs for game_action_t, game_action_condition_t, and game_action_effect_t and functions for adding/getting/checking possible actions for an item.
- Issue #315: Developing functions for the game_action_condition_t struct, including functions to create new conditions, add them to items, and check if conditions are met
- Issue #316: Developing functions for the game_action_effect_t struct, including functions to create new effects, add them to items, and carry out effects
- Issue #331: Adding a hash table of items into the game struct and adding functions to add and retrieve items from the hash table.

Changes that were made in this PR process (many of these changes were stylistic or documentation-related, leading to the big leap in lines of code added):

- Making styling consistent to K&R standards (not Allman)
- Used macros to define error codes and make return values consistent (see game_state_common.h in include)
- Removed commented out code/stderr calls/old tests for obsolete functions
- Improved documentation for functions
- Removed an unnecessary item argument from add_effect's inputs
- Made corrections to mistakes in possible_action, add_effect, add_condition, and last_game_in_room per other teams' requests
- Removed and corrected code related to hidden pointers in our hash table typedefs
- Moved game_action_new() to common-game-action, moved get_attribute into item.c
- Added functions for freeing llists of action_effect and action_condition

lydiafilipe commented:
Issue Score: ✓

Comments:
Nice work! However, this pull request was a bit large.

<!-- {"score": "check"} -->
